crates/walrus-sui/src/client/retry_client.rs

```diff
@@ -287,6 +287,110 @@ where
+    /// The default maximum number of retries.
+    const DEFAULT_MAX_RETRIES: usize = 3;
+
+    /// Creates a new failover wrapper.
+    pub fn new(instances: Vec<(T, String)>) -> anyhow::Result<Self> {
+        if instances.is_empty() {
+            return Err(anyhow::anyhow!("No clients available"));
+        }
+        Ok(Self {
+            max_retries: Self::DEFAULT_MAX_RETRIES.min(instances.len()),
+            inner: Arc::new(
+                instances
+                    .into_iter()
+                    .map(|(client, name)| (Arc::new(client), name))
+                    .collect(),
+            ),
+            current_index: Arc::new(AtomicUsize::new(0)),
+        })
+    }
+
+    fn client_count(&self) -> usize {
+        self.inner.len()
+    }
+
+    /// Gets a client at the specified index (wrapped around if needed).
+    async fn get_client(&self, index: usize) -> Arc<T> {
+        let wrapped_index = index % self.client_count();
+        self.inner[wrapped_index].0.clone()
+    }
+
+    /// Gets the name of the client at the specified index.
+    fn get_name(&self, index: usize) -> &str {
+        &self.inner[index % self.client_count()].1
+    }
+
+    /// Executes an operation on the current inner instance, falling back to the next one
+    /// if it fails.
+    async fn with_failover<F, Fut, R>(&self, operation: F) -> Result<R, RetriableClientError>
+    where
+        F: for<'a> Fn(Arc<T>) -> Fut,
+        Fut: Future<Output = Result<R, RetriableClientError>>,
+    {
+        let mut last_error = None;
+        let start_index = self
+            .current_index
+            .load(std::sync::atomic::Ordering::Relaxed);
+        let mut current_index = start_index;
+
+        for _ in 0..self.max_retries.min(self.client_count()) {
+            let instance = self.get_client(current_index).await;
+
+            let result = {
+                #[cfg(msim)]
+                {
+                    let mut inject_error = false;
+                    fail_point_if!("fallback_client_inject_error", || {
+                        inject_error = should_inject_error(current_index);
+                    });
+                    if inject_error {
+                        Err(RetriableClientError::RpcError(
+                            tonic::Status::internal("injected error for testing").into(),
+                        ))
+                    } else {
+                        operation(instance).await
+                    }
+                }
+                #[cfg(not(msim))]
+                {
+                    operation(instance).await
+                }
+            };
+
+            match result {
+                Ok(result) => {
+                    self.current_index
+                        .store(current_index, std::sync::atomic::Ordering::Relaxed);
+                    return Ok(result);
+                }
+                Err(e) => {
+                    last_error = Some(e);
+                    tracing::warn!(
```

Comment:
This should probably be a debug level, as today we do not log at warn level when retrying the same endpoint (to keep the behavior consistent)